mat64/qr.go

```diff
@@ -7,179 +7,211 @@ package mat64
-// so QR will never fail unless m < n. The primary use of the QR decomposition is
-// in the least squares solution of non-square systems of simultaneous linear equations.
-// This will fail if QRIsFullRank() returns false. The matrix a is overwritten by the
-// decomposition.
-func QR(a *Dense) QRFactor {
-	// Initialize.
+// The QR decomposition is a factorization of the matrix A such that A = Q * R.
+// The matrix Q is an orthonormal m×m matrix, and R is an m×n upper triangular matrix.
+// Q and R can be extracted from the QFromQR and RFromQR methods on Dense.
+func (qr *QR) Factorize(a Matrix) {
 	m, n := a.Dims()
 	if m < n {
 		panic(ErrShape)
 	}
-
-	qr := a
-	rDiag := make([]float64, n)
-
-	// Main loop.
-	for k := 0; k < n; k++ {
-		// Compute 2-norm of k-th column without under/overflow.
-		var norm float64
-		for i := k; i < m; i++ {
-			norm = math.Hypot(norm, qr.at(i, k))
-		}
-
-		if norm != 0 {
-			// Form k-th Householder vector.
-			if qr.at(k, k) < 0 {
-				norm = -norm
-			}
-			for i := k; i < m; i++ {
-				qr.set(i, k, qr.at(i, k)/norm)
-			}
-			qr.set(k, k, qr.at(k, k)+1)
-
-			// Apply transformation to remaining columns.
-			for j := k + 1; j < n; j++ {
-				var s float64
-				for i := k; i < m; i++ {
-					s += qr.at(i, k) * qr.at(i, j)
-				}
-				s /= -qr.at(k, k)
-				for i := k; i < m; i++ {
-					qr.set(i, j, qr.at(i, j)+s*qr.at(i, k))
-				}
-			}
-		}
-		rDiag[k] = -norm
+	k := min(m, n)
+	if qr.qr == nil {
+		qr.qr = &Dense{}
 	}
+	qr.qr.Clone(a)
+	work := make([]float64, 1)
+	qr.tau = make([]float64, k)
+	lapack64.Geqrf(qr.qr.mat, qr.tau, work, -1)
 
-	return QRFactor{qr, rDiag}
+	work = make([]float64, int(work[0]))
+	lapack64.Geqrf(qr.qr.mat, qr.tau, work, len(work))
 }
 
-// IsFullRank returns whether the R matrix and hence a has full rank.
-func (f QRFactor) IsFullRank() bool {
-	for _, v := range f.rDiag {
-		if v == 0 {
-			return false
+// TODO(btracey): Add in the "Reduced" forms for extracting the n×n orthogonal
+// and upper triangular matrices.
+
+// RFromQR extracts the m×n upper trapezoidal matrix from a QR decomposition.
+func (m *Dense) RFromQR(qr *QR) {
+	r, c := qr.qr.Dims()
+	m.reuseAs(r, c)
+	for i := 0; i < r; i++ {
+		for j := 0; j < min(i, c); j++ {
+			m.mat.Data[i*m.mat.Stride+j] = 0
+		}
+		for j := i; j < c; j++ {
+			m.mat.Data[i*m.mat.Stride+j] = qr.qr.mat.Data[i*qr.qr.mat.Stride+j]
 		}
 	}
-	return true
 }
 
-// H returns the Householder vectors in a lower trapezoidal matrix
-// whose columns define the reflections.
-func (f QRFactor) H() *Dense {
-	qr := f.QR
-	m, n := qr.Dims()
-	h := NewDense(m, n, nil)
-	for i := 0; i < m; i++ {
-		for j := 0; j < n; j++ {
-			if i >= j {
-				h.set(i, j, qr.at(i, j))
-			}
+// QFromQR extracts the m×m orthonormal matrix Q from a QR decomposition.
+func (m *Dense) QFromQR(qr *QR) {
+	r, c := qr.qr.Dims()
+	m.reuseAs(r, r)
+
+	// Set Q = I.
+	for i := 0; i < r; i++ {
```

Comment:
This one would actually have to be two steps, one for the top nxn block and then the (m-n)xn block.